File: Problems/25-Flip-String-to-Monotone-Increasing/Flip-String-to-Monotone-Increasing.cpp
class Solution {
public:
    int minFlipsMonoIncr(string s) {
        int n = s.length();
        vector<int> accZero(n+1, 0);
        for (int i = n-1; i>= 0; i--)
        {
            accZero[i] = accZero[i+1];
            if (s[i] == '0') accZero[i]++;
        }

        int ans = accZero[0];
        int numOne = 0;
        for (int i = 0; i < n ; i++)
        {
            if (s[i] == '1') numOne++;
            ans = min(ans, (numOne + accZero[i+1]));
        }
        return ans;
    }
};

File: Problems/25-Flip-String-to-Monotone-Increasing/Flip-String-to-Monotone-Increasing.py
class Solution(object):
    def minFlipsMonoIncr(self, s):
        """
        :type s: str
        :rtype: int
        """
        
        #使得该分界点之前1的个数和分界点之后0的个数之和最小，把分界点之前的1变成0，之后的0变成1
        m=s.count('0')  
        res=[m]
        for x in s:
            if x=='1':  
                m+=1
            else:       
                m-=1
                res.append(m)        
        return min(res)